### PR TITLE
Fixes to ipywidget tests

### LINKED ISSUE #238: IPyWidget tests randomly fail
## Environment data

-   VS Code version: XXX
-   Extension version (available under the Extensions sidebar): XXX
-   Python Extension version (available under the Extensions sidebar): XXX
-   OS (Windows | Mac | Linux distro) and version: XXX
-   Python and/or Anaconda version: XXX
-   Type of virtual environment used (N/A | venv | virtualenv | conda | ...): XXX
-   Jupyter server running: Local | Remote | N/A

## Expected behaviour

XXX

## Actual behaviour

XXX

## Steps to reproduce:

[**NOTE**: Self-contained, minimal reproducing code samples are **extremely** helpful and will expedite addressing your issue]

1. XXX

## Logs

<details>

<summary>Output for <code>Python</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Python</code>)
</summary>

<p>

```
XXX
```

</p>
</details>

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes #238 , #4152
**Found 3 problems:**
* File being overwritten
* Widget file not getting downloaded (we added new code to place js files in a new directory, but didn't update the tests)
* We were looking for output immediately after execution

**Another major issue is:**
* Found that while ipywidget tests were running some other code was still running against the kernel, we were attempting to fetch variables & the like.
* Looks like some code is not completing when we start other tests (dispose isn't working in some places & is messing up with other tests)
* Can't replicate eaily, however have the logs that prove this & seems to have caused failure in the tests (the widgets didn't get a chance to load for some reason at all).

* Also found other instance where some code wasn't disposed & when other tests start, some code is still running (weird, again, not likely the cause, but we have some code thats not getting disposed correctly)

* Insipite of this fix, beakerx seems to fail (failed at least once).
* There might be more flaky code paths (however since we're now properly retrying the tests this might be sufficient, if it fails again, i'll check it out).

### CODE REVIEW

File: src/test/datascience/uiTests/ipywidget.ui.functional.test.ts [hunk 1]
```diff
@@ -408,32 +408,32 @@ use(chaiAsPromised);
...
+                await notebookUI.executeCell(1);
+                const cellOutputHtml = await notebookUI.getCellOutputHTML(1);
+                assert.include(cellOutputHtml, '<canvas ');
+
+                const cellOutput = await notebookUI.getCellOutput(1);
+                const sliders = await cellOutput.$$('div.ui-slider');
+                assert.equal(sliders.length, 2);
+            });
+        });
+        test('Render ipyvolume 2', async function () {
```
[the remark this answers is not in the document]
Comment (DonJayamanne) [reply src/test/datascience/uiTests/ipywidget.ui.functional.test.ts:426]: WebGLs (3d rendering) consume a lot of resources, hence possible CI is constrained or somethign similar - hence the plan to split the two widgets into separate tests (minimize widgets rendered in a single webview).

File: src/client/datascience/ipywidgets/cdnWidgetScriptSourceProvider.ts [hunk 2]
```diff
@@ -111,14 +111,16 @@ export class CDNWidgetScriptSourceProvider implements IWidgetScriptSourceProvide
...
-                traceInfo(
+                traceInfoIf(
                     !!process.env.VSC_JUPYTER_FORCE_LOGGING,
                     `Widget Script downloaded for ${moduleName}:${moduleVersion}, already downloaded ${await this.fs.localFileExists(
                         diskPath
                     )}`
                 );
-                // Need to copy from the temporary file to our real file (note: VSC filesystem fails to copy so just use straight file system)
-                await this.fs.copyLocal(tempFile.filePath, diskPath);
+                if (!(await this.fs.localFileExists(diskPath))) {
```
Comment (DonJayamanne) [inline src/client/datascience/ipywidgets/cdnWidgetScriptSourceProvider.ts:120]: Problem 1 - This was required, else we end up overwriting an existing file

File: src/test/datascience/uiTests/helpers.ts [hunk 3]
```diff
@@ -95,7 +96,9 @@ export class BaseWebUI implements IAsyncDisposable {
             ? setTimeout(async () => {
                   if (!promise.resolved) {
                       // Take a screenshot and save at the root with the same name (upload on error)
-                      await this.page?.screenshot({ path: path.join(EXTENSION_ROOT_DIR, 'browser-screenshot.png') });
+                      await this.page
+                          ?.screenshot({ path: path.join(EXTENSION_ROOT_DIR, 'browser-screenshot.png') })
```
Comment (DonJayamanne) [inline src/test/datascience/uiTests/helpers.ts:100]: Was throwing a lot of unhandled exceptions (page was already disposed at this point)

File: src/test/datascience/uiTests/helpers.ts [hunk 4]
```diff
@@ -141,6 +144,10 @@ export class BaseWebUI implements IAsyncDisposable {
         await this.browser.newContext();
         this.page = await this.browser.newPage();
         await this.page.goto(url);
+        this.page.route('**', (route) => {
+            console.log(`Playwright browser Url requested ${route.request().url()}`);
```
Comment (DonJayamanne) [inline src/test/datascience/uiTests/helpers.ts:148]: Added logging to see what is being requested, as there were a few 404s in browser.

File: src/test/datascience/uiTests/ipywidget.ui.functional.test.ts [hunk 5]
```diff
@@ -34,14 +34,14 @@ use(chaiAsPromised);
 // When using jupyter server, ipywidget tests seem to be a lot flakier. Always use raw kernel
 [true].forEach((useRawKernel) => {
     //import { asyncDump } from '../common/asyncDump';
-    suite(`DataScience IPyWidgets (${useRawKernel ? 'With Direct Kernel' : 'With Jupyter Server'})`, () => {
+    suite(`DataScience IPyWidgets (${useRawKernel ? 'With Direct Kernel' : 'With Jupyter Server'})`, function () {
         const disposables: Disposable[] = [];
         let ioc: DataScienceIocContainer;
+        // These are UI tests, hence nothing to do with platforms.
+        this.timeout(30_000); // UI Tests, need time to start jupyter.
+        this.retries(3); // UI tests can be flaky.
```
Comment (DonJayamanne) [inline src/test/datascience/uiTests/ipywidget.ui.functional.test.ts:42]: None of the ipywidget tests were retrying as this was in the wrong place.
We never retry the tests

File: src/test/datascience/uiTests/webBrowserPanel.ts [hunk 6]
```diff
@@ -186,10 +213,26 @@ export class WebBrowserPanel implements IWebviewPanel, IDisposable {
...
         const filePath = localResource.fsPath;
         const name = path.basename(path.dirname(filePath));
-        if (name !== 'nbextensions' && this.serverUrl) {
-            // This is a CDN download, Remap to our webserver
-            const remapped = `${this.serverUrl}/source?hash=${name}`;
-            return Uri.parse(remapped);
+        if (this.serverUrl) {
+            if (name === 'nbextensions') {
+                // This is a CDN download, Remap to our webserver
+                const remapped = `${this.serverUrl}/nbextensionssource?hash=${path.basename(filePath, '.js')}`;
```
Comment (DonJayamanne) [inline src/test/datascience/uiTests/webBrowserPanel.ts:219]: Problem 3 - These files (in the nbextensions) folder were never getting downloaded. Hence the tests always failed when they resorted to downloaded from tmp/nbextensions folder.

File: src/test/datascience/uiTests/ipywidget.ui.functional.test.ts [hunk 7]
```diff
@@ -408,32 +408,31 @@ use(chaiAsPromised);
...
         });
-        test('Render ipyvolume', async function () {
-            // See bug https://github.com/microsoft/vscode-jupyter/issues/4152
-            this.skip();
-            // const { notebookUI } = await openIPyVolumeIpynb();
-            // await assert.eventually.isFalse(notebookUI.cellHasOutput(3));
-
-            // // Confirm sliders and canvas are rendered.
-            // await retryIfFail(async () => {
-            //     await notebookUI.executeCell(1);
```
Comment (DonJayamanne) [inline src/test/datascience/uiTests/ipywidget.ui.functional.test.ts:419]: Problem 2 - This was in the wrong place.
We execute the cell & immediate look for output, the execution is slow, hence output will not be available majority of the time, hence tests would keep running & creating widgets & then destroying it over & over by executing the cell until the test times out. 
We don't need to re-execute the cells (we can leave that for retrying the test).

**Comment by DonJayamanne** [comment]
Test reports aren't working even when tests pass. Will submit a dummy change.

**Review by joyceerhl — APPROVED** [review]

**Review by rchiodo — APPROVED** [review]

**Pull request merged**